toolkit/components/filewatcher/NativeFileWatcherMac.h

```diff
@@ -0,0 +1,255 @@
+ */
+struct WatchedResourceDescriptor
+{
+    // The path on the file system of the watched resource.
+    nsString mPath;
+    WatchedResourceDescriptor(const nsAString& aPath)
+        : mPath(aPath)
+    {
+    }
+};
+
+/**
+ * The CallBackAction struct is a specific event that occurs in the filesystem as reported by FSEvents.
+ */
+struct CallBackAction{
+    CallBackAction(char*eventPath, const FSEventStreamEventFlags eventFlag, const FSEventStreamEventId eventIds)
+        : mEventFlags(eventFlag)
+        , mEventIds(eventIds)
+    {
+        snprintf(mEventPath, PATH_MAX, "%s", eventPath);
+    }
+
+    char mEventPath[PATH_MAX];
+    const FSEventStreamEventFlags mEventFlags;
+    const FSEventStreamEventId mEventIds;
+};
+
+
+/**
+ * The CallBackEvents struct manages the FSEvents events returned to the stream.
+ *
+ * This is populated inside of the FSEvents callback. It is then processed on a separate
+ * thread by NativeFileWatcherIOTask::RunInternal() which parses the events and calls the
+ * appropriate callbacks.
+ */
+struct CallBackEvents{
+    CallBackEvents()
+        : callBackLock("filewatcher::eventLock")
+    {}
+    std::queue<CallBackAction> mSavedEvents; // Used to hold individual FSEvents
+    Mutex callBackLock; // Used for safe access across multiple threads.
+};
+
+class NativeFileWatcherIOTask; // Forward Declaration of class.
+
+
+/**
+ * The NativeFileWatcherFSETask class which holds all of the functionality of the FSEventStream.
+ *
+ * It inherits from Runnable as we need to make a blocking call to CFRunLoopRun(). CFRunLoopRun() allows the
+ * stream to report file system events. The event stream is an immutable object, therefore with every change in the
+ * paths watched, the stream must be stopped and a new stream must be created with a new list of paths to be watched.
+ */
+class NativeFileWatcherFSETask : public Runnable
+{
+public:
+    explicit NativeFileWatcherFSETask(NativeFileWatcherIOTask* parent, CallBackEvents* cbe, std::vector<CFStringRef>& dirs);
+
+    // This is a blocking method which runs the steam on a CFRunLoop.
+    NS_IMETHOD Run() override;
+
+    // Adds path to the stream (and restarts it)
+    NS_IMETHOD AddPath(char* pathToAdd);
+
+    // Removes path from the stream and restarts it. Unless it removes the last path being watched by the stream.
+    // In this instance, we will return from this runnable without starting a new stream.
+    NS_IMETHOD RemovePath(char* pathToRemove);
+
+private:
+    //mozilla::Mutex runLoopLock;
+
+    FSEventStreamRef mEventStreamRef; // Reference to the FSEventStream
+    CFRunLoopRef mRunLoop = nullptr; // Reference to the RunLoop which is passed back to the parent class, to stop the RunLoop
+    std::vector<CFStringRef> mDirs; // A mutable list of directories which we build dynamically based on the needs of the stream
+    static CallBackEvents* cbe_internal; // Pointer to a callback events struct which is allocated in the parent class and used

```

Comment:
Why is this `static`? Why isn't the other pointer `static`? Why do you need both?

Also, since I don't see a parent class other than `Runnable`, what does this even mean?

Reply:
me to review

Reply:
As in the linux implementation - because it's being used in a static context (the event stream callback). We need a function pointer to the callback when setting up the stream so it's either a static member method (and static variables to use inside of that method), or namespace globals. I'll defer to you on which of those is preferable and we can switch the linux and mac implementations over. And if we leave it as static member methods/variables, I'll make sure we comment why they are static.

Reply:
Either is fine, as long as it's documented.

Reply:
Great, thanks.